script-runner/AutomatedTestRunner.hs

```diff
@@ -0,0 +1,400 @@
+import           Pos.Crypto (emptyPassphrase, hash, hashHexF, withSafeSigners, noPassEncrypt)
+import           Pos.DB.BlockIndex (getTipHeader)
+import           Pos.Util.CompileInfo (CompileTimeInfo (ctiGitRevision), HasCompileInfo, compileInfo, withCompileInfo)
+import           Prelude (show)
+import           Text.PrettyPrint.ANSI.Leijen (Doc)
+import           Universum hiding (when, show, on, state)
+import qualified Pos.Client.CLI as CLI
+import           Pos.Launcher (HasConfigurations, NodeParams (npBehaviorConfig, npUserSecret, npNetworkConfig),
+                     NodeResources, WalletConfiguration,
+                     bracketNodeResources, loggerBracket,
+                     runNode, runRealMode, withConfigurations, InitModeContext)
+import           Formatting (int, sformat, (%), Format)
+import           Graphics.Vty (mkVty, defaultConfig, defAttr)
+import           Ntp.Client (NtpConfiguration)
+import           PocMode (AuxxContext(AuxxContext, acRealModeContext), AuxxMode, realModeToAuxx)
+import           Pos.Chain.Block (LastKnownHeaderTag)
+import           Pos.Chain.Genesis as Genesis (Config (configGeneratedSecrets, configProtocolMagic), configEpochSlots)
+import           Pos.Chain.Txp (TxpConfiguration)
+import           Pos.Chain.Update (UpdateData, SystemTag, mkUpdateProposalWSign, BlockVersion, SoftwareVersion, BlockVersionModifier, updateConfiguration)
+import           Pos.Client.Update.Network (submitUpdateProposal)
+import           Pos.Core (LocalSlotIndex, SlotId (SlotId, siEpoch, siSlot), mkLocalSlotIndex, EpochIndex(EpochIndex), SlotCount, getEpochIndex, getSlotIndex, difficultyL, getChainDifficulty, getBlockCount, getEpochOrSlot)
+import           Pos.DB.DB (initNodeDBs)
+import           Pos.DB.Txp (txpGlobalSettings)
+import           Pos.Infra.Diffusion.Types (Diffusion, hoistDiffusion)
+import           Pos.Infra.Network.Types (NetworkConfig (ncTopology, ncEnqueuePolicy, ncDequeuePolicy, ncFailurePolicy), Topology (TopologyAuxx), topologyDequeuePolicy, topologyEnqueuePolicy, topologyFailurePolicy, NodeId)
+import           Pos.Infra.Shutdown (triggerShutdown)
+import           Pos.Infra.Slotting.Util (onNewSlot, defaultOnNewSlotParams)
+import           Pos.Util (logException, lensOf)
+import           Pos.Util.UserSecret (usVss, readUserSecret, usPrimKey, usKeys)
+import           Pos.Util.Wlog (LoggerName)
+import           Pos.WorkMode (RealMode, EmptyMempoolExt)
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Map as Map
+import qualified Data.Text as T
+
+class TestScript a where
+  getScript :: a -> Script
+
+data ScriptRunnerOptions = ScriptRunnerOptions
+  { srCommonNodeArgs :: !CLI.CommonNodeArgs -- ^ Common CLI args for nodes
+  , srPeers          :: ![NodeId]
+  } deriving Show
+
+data ScriptBuilder = ScriptBuilder
+  { sbScript :: Script
+  , sbEpochSlots :: SlotCount
+  , sbGenesisConfig :: Config
+  }
+data NodeType = Core { ntIdex :: Integer }
```

Comment:
If a data type has a single file, it should be a `newtype` :).

Reply:
it will become a sum type soon, to support relay and maybe wallet nodes